base/setup/lib/utils/filesup.c

```diff
@@ -71,78 +68,84 @@ SetupCreateSingleDirectory(
                           NULL,
                           0);
     if (NT_SUCCESS(Status))
-    {
         NtClose(DirectoryHandle);
-    }
-
-    RtlFreeUnicodeString(&PathName);
 
     return Status;
 }
 
 NTSTATUS
 SetupCreateDirectory(
-    IN PCWSTR PathName)
+    _In_ PCWSTR DirectoryName)
 {
     NTSTATUS Status = STATUS_SUCCESS;
-    PWCHAR PathBuffer = NULL;
+    UNICODE_STRING PathNameU;
+    PWCHAR PathBuffer;
     PWCHAR Ptr, EndPtr;
-    ULONG BackslashCount;
-    ULONG Size;
-
-    Size = (wcslen(PathName) + 1) * sizeof(WCHAR);
-    PathBuffer = RtlAllocateHeap(RtlGetProcessHeap(), HEAP_ZERO_MEMORY, Size);
-    if (PathBuffer == NULL)
-        return STATUS_INSUFFICIENT_RESOURCES;
 
-    wcscpy(PathBuffer, PathName);
-    EndPtr = PathBuffer + wcslen(PathName);
-
-    Ptr = PathBuffer;
+    if (!RtlCreateUnicodeString(&PathNameU, DirectoryName))
+        return STATUS_NO_MEMORY;
+    PathBuffer = PathNameU.Buffer;
+    EndPtr = PathBuffer + (PathNameU.Length / sizeof(WCHAR));
 
-    /* Skip the '\Device\HarddiskX\PartitionY\ part */
-    BackslashCount = 0;
-    while (Ptr < EndPtr && BackslashCount < 4)
+    /* Find the deepest existing sub-directory: start from the
+     * end and go back verifying each sub-directory in turn */
+    for (Ptr = EndPtr; PathBuffer < Ptr;)
     {
-        if (*Ptr == L'\\')
-            BackslashCount++;
-
-        Ptr++;
-    }
+        WCHAR wch = UNICODE_NULL;
+        BOOLEAN bExists;
 
-    while (Ptr < EndPtr)
-    {
-        if (*Ptr == L'\\')
+        /* If we are on a separator, save and erase the next character */
+        if ((Ptr < EndPtr) && (*Ptr == OBJ_NAME_PATH_SEPARATOR))
         {
-            *Ptr = 0;
+            wch = *(Ptr+1);
+            *(Ptr+1) = UNICODE_NULL;
+        }
 
-            DPRINT("PathBuffer: %S\n", PathBuffer);
-            if (!DoesDirExist(NULL, PathBuffer))
-            {
-                DPRINT("Create: %S\n", PathBuffer);
-                Status = SetupCreateSingleDirectory(PathBuffer);
-                if (!NT_SUCCESS(Status))
-                    goto done;
-            }
+        DPRINT("PathBuffer: %S\n", PathBuffer);
+        bExists = DoesDirExist(NULL, PathBuffer);
```

Comment:
Would it not make more sense to have DoesDirExist accept a UNICODE_STRING?
Then you don't have to edit (or even allocate) the buffer.

Reply:
I will consider having a `DoesDirExistU` for that case (although there are many other places where we only have a string buffer (not UNICODE_STRING) where that function gets called to).

Reply:
I've implemented your suggestions in the latest commit, and indeed this simplifies all that.